tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,397 @@
+    do NOT indicate stdin EoF until after some data has been received from the server.
+    """
+    close_marker_bytes = data_bytes(10)
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+    client_options.port = port
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.port = port
+    server_options.key = certificate.key
+    server_options.cert = certificate.cert
+    server_options.data_to_send = close_marker_bytes
+
+    assert not os.path.exists(options.ticket_file)
+
+    s2n_server = managed_process(S2N, server_options, send_marker=S2N.get_send_marker())
+    client = managed_process(provider, client_options, close_marker=str(close_marker_bytes))
+
+    for results in s2n_server.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+
+    for results in client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+
+    assert os.path.exists(options.ticket_file)
+
+
+"""
+Basic S2N server happy case.
+
+We make one full connection to get a session ticket with early data enabled,
+then another resumption connection with early data.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", CLIENT_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("early_data_size", [int(MAX_EARLY_DATA/2), int(MAX_EARLY_DATA-1), MAX_EARLY_DATA, 1])
+def test_s2n_server_with_early_data(managed_process, tmp_path, cipher, curve, protocol, provider, certificate, early_data_size):
+    ticket_file = str(tmp_path / TICKET_FILE)
+    early_data_file = str(tmp_path / EARLY_DATA_FILE)
+    early_data = get_early_data_bytes(early_data_file, early_data_size)
+
+    options = ProviderOptions(
+        port=next(available_ports),
+        cipher=cipher,
+        curve=curve,
+        protocol=protocol,
+        insecure=True,
+        use_session_ticket=True,
+        data_to_send=DATA_TO_SEND,
+    )
+    options.ticket_file = ticket_file
+    options.early_data_file = early_data_file
+    options.max_early_data = MAX_EARLY_DATA
+
+    get_ticket_from_s2n_server(options, managed_process, provider, certificate)
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+
+    s2n_server = managed_process(S2N, server_options)
+    client = managed_process(provider, client_options)
+
+    for results in client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
```

Comment:
Can we assert on the client side/OpenSSL logs that the early data was received by the server?

Reply:
I'm not sure I understand the suggestion. We verify that the server received the early data by checking the server output. Why would the client be a good source for that information?